v3/mathjax3-ts/handlers/html/HTMLMathItem.ts

```diff
@@ -0,0 +1,131 @@
+/*************************************************************
+ *
+ *  Copyright (c) 2017 The MathJax Consortium
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/**
+ * @fileoverview  Implements the HTMLMathItem object
+ *
+ * @author dpvc@mathjax.org (Davide Cervone)
+ */
+
+import {AbstractMathItem, Location} from '../../core/MathItem.js';
+import {InputJax} from '../../core/InputJax.js';
+import {HTMLDocument} from './HTMLDocument.js';
+
+/*****************************************************************/
+/*
+ *  Implements the HTMLMathItem class (extends AbstractMathItem)
+ */
+
+export class HTMLMathItem extends AbstractMathItem {
+
+    public static STATE = AbstractMathItem.STATE;
+
+    /*
+     * @override
+     */
+    constructor(math: string, jax: InputJax, display: boolean = true,
+                start: Location = {node: null, n: 0, delim: ''},
+                end: Location = {node: null, n: 0, delim: ''}) {
+        super(math, jax, display, start, end);
+    }
+
+    /*
+     * Not yet implemented
+     *
+     * @override
+     */
+    public addEventHandlers() {}
+
+    /*
+     * Insert the typeset MathItem into the document at the right location
+     *   If the starting and ending nodes are the same:
+     *     Split the text to isolate the math and its delimiters
+     *     Replace the math by the typeset version
+     *   Otherewise (spread over several nodes)
+     *     Split the start node, if needed
+     *     Remove nodes until we reach the end node
+     *     Insert the math before the end node
+     *     Split the end node, if needed
+     *     Remove the end node
+     *
+     * @override
+     */
+    public updateDocument(html: HTMLDocument) {
+        if (this.state() < STATE.INSERTED) {
+            if (this.inputJax.processStrings) {
+                let node = this.start.node as Text;
+                if (node === this.end.node) {
+                    if (this.end.n < this.end.node.nodeValue.length) {
+                        this.end.node.splitText(this.end.n);
+                    }
+                    if (this.start.n) {
+                        node = (this.start.node as Text).splitText(this.start.n);
```

Comment:
No, `this.start.node` is the node in which the start of the math expression occurs, not the start of the document.  The math may span several nodes (e.g., several Text nodes, or two text nodes separated by ` <br>` node), so we may need to traverse several nodes to get the text of the math.  We split the node where the start occurs (at the start location, which is `this.start.n`, and use the second half as the beginning of the math.  Then collect text nodes until we get to the `this.end.node`, and split that to get the final piece of the text.

Note that we aren't searching through the document at this point, as we have pointers to the nodes already (the searching was done earlier with `findMath()`).